lib/puppet/util/windows/taskscheduler2.rb

```diff
@@ -0,0 +1,707 @@
+  #
+  # DONE
+  def enum_task_names(folder_path = ROOT_FOLDER, options = nil)
+    raise Error.new(_('No current task scheduler. Schedule.Service is NULL.')) if @task_service.nil?
+    raise TypeError unless folder_path.is_a?(String)
+
+    options = {} if options.nil?
+    options[:include_child_folders] = true if options[:include_child_folders].nil?
+    options[:include_compatibility] = [] if options[:include_compatibility].nil?
+
+    array = []
+
+    task_folder = @task_service.GetFolder(folder_path)
+
+    task_folder.GetTasks(TASK_ENUM_HIDDEN).each do |task|
+      included = true
+
+      unless !included || options[:include_compatibility].empty?
+        included = included && options[:include_compatibility].include?(task.Definition.Settings.Compatibility)
+      end
+
+      array << task.Path if included
+    end
+    return array unless options[:include_child_folders]
+
+    task_folder.GetFolders(0).each do |child_folder|
+      array = array + enum_task_names(child_folder.Path, options)
+    end
+
+    array
+  end
+
+  # DONE
+  def activate(task)
+    raise Error.new(_('No current task scheduler. Schedule.Service is NULL.')) if @task_service.nil?
+    raise TypeError unless task.is_a?(String)
+
+    task_folder = @task_service.GetFolder(get_folder_path_from_task(task))
+
+    begin
+      @pITask = task_folder.GetTask(get_task_name_from_task(task))
+    rescue WIN32OLERuntimeError => e
+      @pITask = nil
+      # TODO win32ole errors are horrible.  Assume the task doesn't exist
+    end
+    @pITaskDefinition = nil
+    
+    @pITask
+  end
+
+  # DONE
+  def deactivate()
+    @pITask = nil
+    @pITaskDefinition = nil
+  end
+
+  # DONE
+  def definition()
+    if @pITaskDefinition.nil? && !@pITask.nil?
+      # Create a new editable Task Defintion based off of the currently activated task
+      @pITaskDefinition = @task_service.NewTask(0)
+      @pITaskDefinition.XmlText = @pITask.XML
+    end
+
+    @pITaskDefinition
+  end
+
+  # Delete the specified task name.
+  #
+  # DONE
+  def delete(task)
+    raise Error.new(_('No current task scheduler. Schedule.Service is NULL.')) if @task_service.nil?
+    raise TypeError unless task.is_a?(String)
+
+    task_folder = @task_service.GetFolder(get_folder_path_from_task(task))
+
+    result = -1
+    begin
+      result = task_folder.DeleteTask(get_task_name_from_task(task),0)
+    rescue WIN32OLERuntimeError => e
+      # TODO win32ole errors are horrible.  Assume the task doesn't exist so deletion is successful
+      return true
+    end
+
+    result == S_OK
+  end
+
+  # Execute the current task.
+  #
+  # DONE
+  def run(arguments = nil)
+    raise Error.new(_('No currently active task. ITask is NULL.')) if @pITask.nil?
+
+    @pITask.Run(arguments)
+  end
+
+  # Saves the current task. Tasks must be saved before they can be activated.
+  #
+  # DONE
+  def save()
```

Comment:
Something to think about for a future commit / iteration of the API - do we want to pass in the `definition` and/or the `task`?